### ISSUE TITLE
New world spawn has far too long a road

### ISSUE THREAD

**Issue by ilovehans10** [CleverRaven/Cataclysm-DDA#67427]
### Describe the bug

With the new feature (#66781) that reveals the map leading to the center of the nearest town (which I absolutely love) I have just run into a world where on spawn there was a long road. It looks like this was caused by there being no easy way to draw a road from the spawn to the nearest town, so there wasn't one made. I also thought of #64717 when I saw the bug, but that might just be because it is late and I am tired.

### Attach save file

[Debug-trimmed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/12256507/Debug-trimmed.tar.gz)

### Steps to reproduce

I just hit play now as this was supposed to be my world to test my pull #67417, but on spawn I saw the weird map when I opened it on habit.

Seriously though, I feel like this is a nearly impossible thing to reproduce, but I have the save to prove it happened.

### Expected behavior

I would say with the new roads being drawn shorter between towns, it might be a good idea to have the town with the shortest road from player spawn selected instead of the closest town.

### Screenshots

![Screenshot from 2023-08-03 21-17-48](https://github.com/CleverRaven/Cataclysm-DDA/assets/20823837/c9b7e31d-4cb7-4dc5-8f32-731c86adeb6d)
![Screenshot from 2023-08-03 21-17-54](https://github.com/CleverRaven/Cataclysm-DDA/assets/20823837/61a66ae6-4ef5-4a8b-86a5-86f36b14c8c5)

### Versions and configuration

- OS: Linux
    - OS Version: LSB Version: n/a; Distributor ID: Arch; Description: Arch Linux; Release: rolling; Codename: n/a; 
- Game Version: 0.G-3439-gc5ecac002f-dirty [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth],
    Stats Through Skills [StatsThroughSkills],
    Stats Through Kills [stats_through_kills],
    SpeedyDex [speedydex],
    Bionic Professions [package_bionic_professions],
    Aftershock [aftershock],
    Blaze Industries [blazeindustries],
    Magiclysm [magiclysm],
    Extra Mutated Scenarios [extra_mut_scens],
    sees-player, run, crouch, bleed icons, retrodays [sees_player_retro]
]

### Additional context

I know it says my game version is dirty, but I am almost certain that that is only because I added 4 words to a string in src/character.cpp which shouldn't affect world gen at all.

**Comment by GlordXd** [comment]
The game just went like: no,lets make things a lil bit complicated,no need to go to the closest town

But i feel like its might be intended like this,just _very_ rare (story-telling potential?)

**Comment by ilovehans10** [comment]
> It probably shoud instead make list of some nearby cities, calculate how long the road to them, and pick the shortest one, but i have no clue is it doable

Would it be possible to make it so that the game "traverses" all the roads that are within a specific radius of spawn until it either reaches a city or hits a limit based on the city spacing. This would make it so that if you don't spawn on a road it will still give you a little help, and work well with the new way that roads generate. Also, I am relatively new to contributing, is this something where we should at RenechCDDA as this is probably related to their pull request, or did they get notified when I mentioned their pull request?

**Comment by Brambor** [comment]
Yep, BFS until you reach a city center, then pick road to that city.

**Comment by sparr** [comment]
I agree that the closest-by-travel-distance path should be used instead of closest-as-the-crow-flies, at least by default.

BFS would work, but here's another idea:

Pick the closest city as the crow flies and path to it, same as now. But as soon as that path enters another city, change the destination to the center of that city.

**Comment by RenechCDDA** [comment]
> I agree that the closest-by-travel-distance path should be used instead of closest-as-the-crow-flies, at least by default.
> 
> BFS would work, but here's another idea:
> 
> Pick the closest city as the crow flies and path to it, same as now. But as soon as that path enters another city, change the destination to the center of that city.

Basically everything in this comment is the way to go. Low priority fix, but it would be nice to have.
